lib/puppet/pops/loader/runtime3_type_loader.rb

```diff
@@ -21,28 +22,51 @@ def to_s()
   # @return [Loader::NamedEntry, nil found/created entry, or nil if not found
   #
   def find(typed_name)
-    return nil unless typed_name.name_authority == Pcore::RUNTIME_NAME_AUTHORITY && typed_name.type == :type
+    return nil unless typed_name.name_authority == Pcore::RUNTIME_NAME_AUTHORITY
+    case typed_name.type
+    when :type
+      impl_te = find_impl(TypedName.new(:resource_type_pp, typed_name.name, typed_name.name_authority))
+      if impl_te.nil? || impl_te.value.nil?
+        nil
+      else
+        # Loaded types doesn't have the same life cycle as this loader, so we must start by
+        # checking if the type was created. If it was, an entry will already be stored in
+        # this loader. If not, then it was created before this loader was instantiated and
+        # we must therefore add it.
+        te = get_entry(typed_name)
+        te = set_entry(typed_name, Types::TypeFactory.resource(impl_te.value.name.to_s)) if te.nil? || te.value.nil?
+        te
+      end
+    when :resource_type_pp
+      find_impl(typed_name)
+    else
+      nil
+    end
+  end
 
+  # Find the implementation for the resource type by first consulting the internal loader for pp defined 'Puppet::Resource::ResourceType3'
+  # instances, then check for a Puppet::Type and lastly check for a defined type.
+  #
+  def find_impl(typed_name)
     name = typed_name.name
-    value = @environment.known_resource_types.find_definition(name)
-    if value.nil?
+    te = Types::TypeParser.type_map.include?(name) ? nil : @resource_3x_loader.find(typed_name)
+    if te.nil? || te.value.nil?
       # Look for Puppet::Type
       value = Puppet::Type.type(name)
```

Comment:
this step can be skipped if the name is qualified since no Puppet::Type ever finds a qualified name (there may be many thousands such calls)